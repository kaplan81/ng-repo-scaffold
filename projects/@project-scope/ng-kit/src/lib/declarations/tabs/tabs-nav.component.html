<nav class="kit-tabs__header" role="tablist">
  <a
    class="kit-tabs__label"
    role="tab"
    *ngFor="let nav of navs"
    [routerLink]="nav.path"
    routerLinkActive="kit-active"
    [innerHTML]="label(nav)">
  </a>
</nav>

<div class="kit-tabs__content">
  <router-outlet></router-outlet>
</div>
