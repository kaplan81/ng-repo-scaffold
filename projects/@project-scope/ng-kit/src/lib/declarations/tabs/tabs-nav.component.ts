import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { NavLink } from '../../models/nav.model';

@Component({
  selector: 'kit-tabs-nav',
  templateUrl: './tabs-nav.component.html',
  styleUrls: ['./tabs.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None
})
export class KitTabsNavComponent {
  @Input()
  navs: NavLink<any>[];

  constructor() {}

  label(nav: NavLink<any>): string {
    return nav.icon
      ? `<i class="material-icons nav-tab-icon">${nav.icon}</i>${nav.text}`
      : nav.text;
  }
}
