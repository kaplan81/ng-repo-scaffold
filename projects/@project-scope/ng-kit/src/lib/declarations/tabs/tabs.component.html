<div class="kit-tabs__header" role="tablist">
  <div class="kit-tabs__label" role="tab" *ngFor="let tab of tabs; let i = index" [class.kit-active]="selectedIndex === i"
    (click)="changeSelectedIndex(i)" [innerHTML]="tab.label">
  </div>
</div>
<div class="kit-tabs__content" role="tabpanel">
  <ng-container *ngFor="let tab of tabs; let i = index">
    <kit-tab-content *ngIf="selectedIndex === i" [content]="tab.content"></kit-tab-content>
  </ng-container>
</div>
