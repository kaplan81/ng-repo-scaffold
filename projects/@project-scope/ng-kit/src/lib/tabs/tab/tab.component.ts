import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { TemplatePortal } from '@angular/cdk/portal';
import { ChangeDetectionStrategy, Component, ContentChild, Input, OnInit, TemplateRef, ViewChild, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { KitTabContentDirective } from '../tab-content/tab-content.directive';

@Component({
  selector: 'kit-tab',
  templateUrl: './tab.component.html',
  styleUrls: ['./tab.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None
})
export class KitTabComponent implements OnInit {
  @Input()
  get isActive(): any {
    return this._isActive;
  }
  set isActive(value: any) {
    this._isActive = coerceBooleanProperty(value);
  }
  @Input()
  label = '';
  @ContentChild(KitTabContentDirective, { read: TemplateRef })
  templateContent: TemplateRef<any>;
  @ViewChild(TemplateRef)
  plainContent: TemplateRef<any>;

  private _isActive = false;
  private contentPortal: TemplatePortal;

  constructor(private viewContainerRef: ViewContainerRef) {}

  ngOnInit(): void {
    this.contentPortal = new TemplatePortal(
      this.templateContent || this.plainContent,
      this.viewContainerRef
    );
  }

  get content(): TemplatePortal {
    return this.contentPortal;
  }
}
