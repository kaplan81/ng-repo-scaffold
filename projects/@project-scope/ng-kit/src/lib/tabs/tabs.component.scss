kit-tabs {
  display: flex;
  flex-direction: column;
  .kit-tabs__header {
    display: flex;
    // Provisional
    border-bottom: 1px solid black;
    .kit-tabs__label {
      display: inline-flex;
      justify-content: center;
      align-items: center;
      text-align: center;
      white-space: nowrap;
      min-width: 160px;
      height: 48px;
      box-sizing: border-box;
      padding: 0 24px;
      color: #fff;
      cursor: pointer;
      // Provisional
      border-top: 1px solid black;
      border-left: 1px solid black;
      border-right: 1px solid black;
      background-color: #347ca2;
      &.kit-active {
        text-shadow: 1px 1px #ccc2b4;
      }
    }
  }
  .kit-tabs__content {
    display: flex;
    flex-grow: 1;
    position: relative;
    overflow: hidden;
    padding: 24px;
    border-bottom: 1px solid black;
    border-left: 1px solid black;
    border-right: 1px solid black;
    background-color: #f2f2f2;
  }
}



