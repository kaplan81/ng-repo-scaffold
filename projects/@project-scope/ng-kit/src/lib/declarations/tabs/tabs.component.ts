import {
  AfterContentInit,
  ChangeDetectionStrategy,
  Component,
  ContentChildren,
  EventEmitter,
  Output,
  QueryList,
  ViewEncapsulation
} from '@angular/core';
import { KitTabComponent } from './tab/tab.component';

@Component({
  selector: 'kit-tabs',
  templateUrl: './tabs.component.html',
  styleUrls: ['./tabs.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None
})
export class KitTabsComponent implements AfterContentInit {
  @ContentChildren(KitTabComponent)
  tabs: QueryList<KitTabComponent>;

  @Output()
  readonly selectedIndexChange: EventEmitter<number> = new EventEmitter<number>();

  selectedIndex = 0;

  ngAfterContentInit() {
    this.tabs.forEach((tab: KitTabComponent, i: number) => {
      if (tab.isActive) this.selectedIndex = i;
    });
  }

  changeSelectedIndex(i: number): void {
    if (this.selectedIndex !== i) {
      this.selectedIndex = i;
      this.selectedIndexChange.emit(this.selectedIndex);
    }
  }
}
