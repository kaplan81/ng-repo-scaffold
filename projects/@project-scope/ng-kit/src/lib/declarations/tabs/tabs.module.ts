import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { KitTabContentComponent } from './tab-content/tab-content.component';
import { KitTabContentDirective } from './tab-content/tab-content.directive';
import { KitTabComponent } from './tab/tab.component';
import { KitTabsNavComponent } from './tabs-nav.component';
import { KitTabsComponent } from './tabs.component';

@NgModule({
  declarations: [
    KitTabsComponent,
    KitTabsNavComponent,
    KitTabComponent,
    KitTabContentComponent,
    KitTabContentDirective
  ],
  exports: [
    RouterModule,
    KitTabsComponent,
    KitTabsNavComponent,
    KitTabComponent,
    KitTabContentDirective
  ],
  imports: [CommonModule, RouterModule, PortalModule]
})
export class KitTabsModule {}
