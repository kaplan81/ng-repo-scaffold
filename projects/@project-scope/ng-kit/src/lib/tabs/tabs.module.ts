import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { KitTabContentComponent } from './tab-content/tab-content.component';
import { KitTabContentDirective } from './tab-content/tab-content.directive';
import { KitTabComponent } from './tab/tab.component';
import { KitTabsComponent } from './tabs.component';

@NgModule({
  declarations: [KitTabsComponent, KitTabComponent, KitTabContentComponent, KitTabContentDirective],
  exports: [KitTabsComponent, KitTabComponent, KitTabContentDirective],
  imports: [CommonModule, PortalModule]
})
export class KitTabsModule {}
