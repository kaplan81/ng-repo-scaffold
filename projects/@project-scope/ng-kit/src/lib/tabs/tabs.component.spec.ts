/* tslint:disable:no-unused-variable */
import { Component, DebugElement } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { click, ComponentSuite, ComponentSuiteElements } from '@project-scope/test-kit';
import { KitTabComponent } from './tab/tab.component';
import { KitTabsComponent } from './tabs.component';
import { KitTabsModule } from './tabs.module';

// bascic case
@Component({
  template: `
    <kit-tabs>
      <kit-tab label="Label 1">Content 1</kit-tab>
      <kit-tab label="Label 2">Content 2</kit-tab>
      <kit-tab label="Label 3">Content 3</kit-tab>
    </kit-tabs>
  `
})
class BasicTabsTestApp {}

// dynamic case
@Component({
  template: `
    <kit-tabs>
      <kit-tab *ngFor="let tab of tabs" label="tab.label">tab.content</kit-tab>
    </kit-tabs>
  `
})
class DynamicTabsTestApp {
  tabs = [
    { label: 'Label 1', content: 'Content 1' },
    { label: 'Label 2', content: 'Content 2' },
    { label: 'Label 3', content: 'Content 3' }
  ];
}

// with active tab
@Component({
  template: `
    <kit-tabs>
      <kit-tab label="Label 1">Content 1</kit-tab>
      <kit-tab label="Label 2" isActive>Content 2</kit-tab>
      <kit-tab label="Label 3">Content 3</kit-tab>
    </kit-tabs>
  `
})
class IsActiveTabsTestApp {}

// with selectedIndexChange event binding
@Component({
  template: `
    <kit-tabs (selectedIndexChange)="selectIndex($event)">
      <kit-tab label="Label 1">Content 1</kit-tab>
      <kit-tab label="Label 2">Content 2</kit-tab>
      <kit-tab label="Label 3">Content 3</kit-tab>
    </kit-tabs>
  `
})
class SelectedEventTabsTestApp {
  index: number = null;

  selectIndex(index: number): void {
    this.index = index;
  }
}

// with kitTabContent directive
@Component({
  template: `
    <kit-tabs>
      <kit-tab label="Label 1">{{plainContent}}</kit-tab>
      <kit-tab label="Label 2">
        <ng-template kitTabContent>
          {{templateContent}}
        </ng-template>
      </kit-tab>
      <kit-tab label="Label 3">Content 3</kit-tab>
    </kit-tabs>
  `
})
class ContentDirectiveTabsTestApp {
  plainContent = 'plain transcluded content!';
  templateContent = 'kitTabContent directive content!';
}

describe('KitTabsComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [KitTabsModule],
      declarations: [
        BasicTabsTestApp,
        DynamicTabsTestApp,
        IsActiveTabsTestApp,
        SelectedEventTabsTestApp,
        ContentDirectiveTabsTestApp
      ]
    }).compileComponents();
  }));

  describe('basic case', () => {
    let fixture: ComponentFixture<BasicTabsTestApp>;
    let els: ComponentSuiteElements<BasicTabsTestApp, KitTabsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(BasicTabsTestApp);
      els = new ComponentSuite<BasicTabsTestApp, KitTabsComponent>(fixture, 'kit-tabs').elements;
    });

    it('should match snapshot', () => {
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    it('should generate as many labels as <kit-tab> tags are in the template', () => {
      fixture.detectChanges();

      const labels: number = els.nested.debugEl.queryAll(By.css('.kit-tabs__label')).length;
      const tabs: number = els.nested.component.tabs.toArray().length;

      expect(labels).toEqual(tabs);
    });

    it('should default to the first tab', () => {
      fixture.detectChanges();

      expect(els.nested.component.selectedIndex).toEqual(0);
    });

    it('should display only the active content', () => {
      fixture.detectChanges();

      const contents: number = els.nested.debugEl.queryAll(By.css('kit-tab-content')).length;

      expect(contents).toEqual(1);
    });

    it('should change selected index on click', () => {
      fixture.detectChanges();

      const secondLabel: DebugElement = els.nested.debugEl.queryAll(By.css('.kit-tabs__label'))[1];
      secondLabel.nativeElement.click();

      expect(els.nested.component.selectedIndex).toEqual(1);
    });
  });

  describe('dynamic case', () => {
    let fixture: ComponentFixture<DynamicTabsTestApp>;
    let els: ComponentSuiteElements<DynamicTabsTestApp, KitTabsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(DynamicTabsTestApp);
      els = new ComponentSuite<DynamicTabsTestApp, KitTabsComponent>(fixture, 'kit-tabs').elements;
    });

    it('should match snapshot', () => {
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    it('should generate as many labels as <kit-tab> tags are in the template', () => {
      fixture.detectChanges();

      const labels: number = els.nested.debugEl.queryAll(By.css('.kit-tabs__label')).length;
      const tabs: number = els.nested.component.tabs.toArray().length;

      expect(labels).toEqual(tabs);
    });
  });

  describe('with active tab', () => {
    let fixture: ComponentFixture<IsActiveTabsTestApp>;
    let els: ComponentSuiteElements<IsActiveTabsTestApp, KitTabsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(IsActiveTabsTestApp);
      els = new ComponentSuite<IsActiveTabsTestApp, KitTabsComponent>(fixture, 'kit-tabs').elements;
    });

    it('should match snapshot', () => {
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    it('should default to the another tab if isActive property is set on <kit-tab>', () => {
      fixture.detectChanges();

      const tabs: KitTabComponent[] = els.nested.component.tabs.toArray();
      const activeIndex: number = tabs.reverse().findIndex((tab: KitTabComponent) => tab.isActive);

      expect(els.nested.component.selectedIndex).toEqual(activeIndex);
    });
  });

  describe('with selectedIndexChange event binding', () => {
    let fixture: ComponentFixture<SelectedEventTabsTestApp>;
    let els: ComponentSuiteElements<SelectedEventTabsTestApp, KitTabsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(SelectedEventTabsTestApp);
      els = new ComponentSuite<SelectedEventTabsTestApp, KitTabsComponent>(fixture, 'kit-tabs')
        .elements;
    });

    it('should match snapshot', () => {
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    it('should select new index and emit it to the parent', () => {
      fixture.detectChanges();
      // This should be 0.
      const firstSelectedIndex: number = els.nested.component.selectedIndex;
      const labels: DebugElement[] = els.nested.debugEl.queryAll(By.css('.kit-tabs__label'));
      const newIndex = 1;
      // This changes the index
      click(labels[newIndex]);
      // This should be 1.
      const secondSelectedIndex: number = els.nested.component.selectedIndex;
      const parentIndex: number = els.host.component.index;

      expect(firstSelectedIndex).not.toEqual(secondSelectedIndex);
      expect(secondSelectedIndex).toEqual(newIndex);
      expect(parentIndex).toEqual(secondSelectedIndex);
    });
  });

  describe('with kitTabContent directive', () => {
    let fixture: ComponentFixture<ContentDirectiveTabsTestApp>;
    let els: ComponentSuiteElements<ContentDirectiveTabsTestApp, KitTabsComponent>;

    beforeEach(() => {
      fixture = TestBed.createComponent(ContentDirectiveTabsTestApp);
      els = new ComponentSuite<ContentDirectiveTabsTestApp, KitTabsComponent>(fixture, 'kit-tabs')
        .elements;
    });

    it('should match snapshot', () => {
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    it('should display explicit content if tab contains an <ng-template> with kitTabContent directive', () => {
      fixture.detectChanges();

      // First tab contains plain transcluded content and is selected by default.
      const plainContent: string = els.host.component.plainContent;
      const firstTabContent: DebugElement = els.nested.debugEl.query(By.css('kit-tab-content'));
      const firstTabContentHtml: string = firstTabContent.nativeElement.innerHTML;

      expect(firstTabContentHtml).toMatch(plainContent);

      const labels: DebugElement[] = els.nested.debugEl.queryAll(By.css('.kit-tabs__label'));
      // // This changes the index, selects the second tab.
      click(labels[1]);
      fixture.detectChanges();
      // Second tab contains an <ng-template> with kitTabContent directive
      // and it is not selected by default.
      const templateContent: string = els.host.component.templateContent;
      const secondTabContent: DebugElement = els.nested.debugEl.query(By.css('kit-tab-content'));
      const secondTabContentHtml: string = secondTabContent.nativeElement.innerHTML;

      expect(secondTabContentHtml).toMatch(templateContent);
    });
  });
});
