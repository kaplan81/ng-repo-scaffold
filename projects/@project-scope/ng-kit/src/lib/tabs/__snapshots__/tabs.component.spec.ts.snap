// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KitTabsComponent basic case should match snapshot 1`] = `
<ng-component>
  <kit-tabs>
    <div
      class="kit-tabs__header"
      role="tablist"
    >
      
      <div
        class="kit-tabs__label kit-active"
        role="tab"
      >
         Label 1 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 2 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 3 
      </div>
    </div>
    <div
      class="kit-tabs__content"
      role="tabpanel"
    >
      
      
      
      <kit-tab-content
        ng-reflect-content="[object Object]"
      >
        
        Content 1
        
      </kit-tab-content>
      
      
      
      
      
      
      
    </div>
  </kit-tabs>
</ng-component>
`;

exports[`KitTabsComponent dynamic case should match snapshot 1`] = `
<ng-component
  tabs={[Function Array]}
>
  <kit-tabs>
    <div
      class="kit-tabs__header"
      role="tablist"
    >
      
      <div
        class="kit-tabs__label kit-active"
        role="tab"
      >
         tab.label 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         tab.label 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         tab.label 
      </div>
    </div>
    <div
      class="kit-tabs__content"
      role="tabpanel"
    >
      
      
      
      <kit-tab-content
        ng-reflect-content="[object Object]"
      >
        
        tab.content
        
      </kit-tab-content>
      
      
      
      
      
      
      
    </div>
  </kit-tabs>
</ng-component>
`;

exports[`KitTabsComponent with active tab should match snapshot 1`] = `
<ng-component>
  <kit-tabs>
    <div
      class="kit-tabs__header"
      role="tablist"
    >
      
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 1 
      </div>
      <div
        class="kit-tabs__label kit-active"
        role="tab"
      >
         Label 2 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 3 
      </div>
    </div>
    <div
      class="kit-tabs__content"
      role="tabpanel"
    >
      
      
      
      
      
      
      <kit-tab-content
        ng-reflect-content="[object Object]"
      >
        
        Content 2
        
      </kit-tab-content>
      
      
      
      
    </div>
  </kit-tabs>
</ng-component>
`;

exports[`KitTabsComponent with kitTabContent directive should match snapshot 1`] = `
<ng-component
  plainContent={[Function String]}
  templateContent={[Function String]}
>
  <kit-tabs>
    <div
      class="kit-tabs__header"
      role="tablist"
    >
      
      <div
        class="kit-tabs__label kit-active"
        role="tab"
      >
         Label 1 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 2 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 3 
      </div>
    </div>
    <div
      class="kit-tabs__content"
      role="tabpanel"
    >
      
      
      
      <kit-tab-content
        ng-reflect-content="[object Object]"
      >
        
        plain transcluded content!
        
      </kit-tab-content>
      
      
      
      
      
      
      
    </div>
  </kit-tabs>
</ng-component>
`;

exports[`KitTabsComponent with selectedIndexChange event binding should match snapshot 1`] = `
<ng-component
  index="null"
>
  <kit-tabs>
    <div
      class="kit-tabs__header"
      role="tablist"
    >
      
      <div
        class="kit-tabs__label kit-active"
        role="tab"
      >
         Label 1 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 2 
      </div>
      <div
        class="kit-tabs__label"
        role="tab"
      >
         Label 3 
      </div>
    </div>
    <div
      class="kit-tabs__content"
      role="tabpanel"
    >
      
      
      
      <kit-tab-content
        ng-reflect-content="[object Object]"
      >
        
        Content 1
        
      </kit-tab-content>
      
      
      
      
      
      
      
    </div>
  </kit-tabs>
</ng-component>
`;
